offsetquery.go

```diff
@@ -0,0 +1,258 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"golang.org/x/sync/errgroup"
+)
+
+// offsetQueryer knows how to send bulk offset queries to Kafka.
+type offsetQueryer struct {
+	topic    string
+	client   sarama.Client
+	consumer sarama.Consumer
+}
+
+// runQuery runs the given offset query and puts the results into info.
+func (r *offsetQueryer) runQuery(ctx context.Context, q *offsetQuery, info *offsetInfo) error {
+	egroup, ctx := errgroup.WithContext(ctx)
+	egroup.Go(func() error {
+		return r.getTimes(ctx, q.timeQuery, info)
+	})
+	egroup.Go(func() error {
+		return r.getOffsets(ctx, q.offsetQuery, info)
+	})
+	egroup.Go(func() error {
+		return r.getResumes(ctx, q.resumeQuery, info)
+	})
+	if err := egroup.Wait(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *offsetQueryer) getTimes(ctx context.Context, q map[int32]map[int64]bool, info *offsetInfo) error {
+	type result struct {
+		partition int32
+		offset    int64
+		time      time.Time
+	}
+	resultc := make(chan result)
+	nrequests := 0
+	egroup, ctx := errgroup.WithContext(ctx)
+	for p, offsets := range q {
+		p := p
+		// getOffset should never fail because getTime should have ensured
+		// that the newest-offset info is present. If it does, it'll panic before
+		// it returns due to assigning to the new query value.
+		newestOffset, _ := info.getOffset(p, symbolicOffsetRequest(sarama.OffsetNewest), nil)
+		for offset := range offsets {
+			offset := offset
+			if offset < 0 {
+				panic("symbolic offset passed to getTime")
+			}
+			nrequests++
+			egroup.Go(func() error {
+				consumeOffset := offset
+				if consumeOffset == newestOffset && newestOffset > 0 {
+					// We're asking about the time of the newest offset. Although the newest
+					// offset is one beyond the last message, we make a special case
+					// for this, as finding the time of the last message is common.
+					consumeOffset--
+				}
+				if consumeOffset >= newestOffset {
+					// No way to know what the timestamp is, so send a zero timestamp
+					// to signify that fact.
+					resultc <- result{
+						partition: p,
+						offset:    offset,
+					}
+					return nil
+				}
+				// TODO batch fetch requests.
+				mt, err := fetchOneMessageTimestamp(r.client, r.topic, p, consumeOffset)
+				if err != nil {
+					return err
+				}
+				resultc <- result{
+					partition: p,
+					offset:    offset,
+					time:      mt,
+				}
+				return nil
+			})
+		}
+	}
+	egroup.Go(func() error {
+		for i := 0; i < nrequests; i++ {
+			select {
+			case result := <-resultc:
+				info.setTime(result.partition, result.offset, result.time)
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+		return nil
+	})
+	return egroup.Wait()
+}
+
+func (r *offsetQueryer) getResumes(ctx context.Context, q map[int32]bool, info *offsetInfo) error {
+	if len(q) == 0 {
+		return nil
+	}
+	return fmt.Errorf("not yet implemented")
+	//	for p := range q {
```

Comment:
I'm not fond of left-over, commented code.  Is this remedied in a follow-up branch?

Reply:
It will be remedied. It's not left-over - it's a sketch for the future.

Reply:
:+1: